src/Domain/FunctionsOfTime/PiecewisePolynomial.cpp

```diff
@@ -157,8 +158,30 @@ void PiecewisePolynomial<MaxDeriv>::update(
   // during this call, there should never be two different threads trying to
   // update this object at the same time in our algorithm. An update should come
   // from only one place and nowhere else so this emplacement should be ok
+  //
+  // The order of these next three changes are important to maintain
+  // thread-safety and avoid a race condition.
+  //
+  //  1. Add the updated deriv info to the back of the list, keeping the valid
+  //     size of the list to use and the expiration time the same.
+  //  2. Update the valid size to match the new size of the list, keeping the
+  //     expiration time the same.
+  //  3. Update the expiration time
+  //
+  // The expiration time is exposed via `time_bounds` and used to determine
+  // whether it's safe to call the `func`s at a specific time. If we update the
+  // expiration time first, then this class will be in an invalid state because
+  // we have not yet updated the stored values; we have the possibility of using
+  // old values at a time past the old expiration time on one thread, but new
+  // values at a time past the old expiration time on a different thread. Thus
+  // we must update the expiration time last.
+  //
+  // We also want to ensure that we are accessing only valid deriv infos. Thus
+  // we need to update the valid size *after* we emplace the updated deriv info
+  // to ensure everything is consistent. This gives us the ordering above.
   deriv_info_at_update_times_.emplace_back(time_of_update, std::move(func));
   deriv_info_size_.fetch_add(1, std::memory_order_acq_rel);
+  expiration_time_.store(next_expiration_time, std::memory_order_release);
```

Comment:
To be extra safe that there isn't a race condition, you can do a https://en.cppreference.com/w/cpp/atomic/atomic/compare_exchange I _think_ there are still edge cases where there would be a race condition, but it at least adds another check.